### ISSUE TITLE
Support for writing binary data to a logger

### ISSUE THREAD

**Issue by rob-p** [gabime/spdlog#643]
Hi @gabime,

  Thank you for this fantastic library!  We've been using it extensively in our lab's projects, and it's been awesome.  I've been particularly surprised with how versatile `spdlog` is; as it allows us to quite cleanly separate the concerns of what we want to write out from _how_ we want to write it (e.g. [here](https://github.com/COMBINE-lab/pufferfish/blob/develop/src/PufferfishAligner.cpp#L1139), where we generalize over stdout vs. a file and uncompressed vs. compressed).  We're particularly fond of using `spdlog` as a means to sync up output of messages from many different threads to a file in a threadsafe manner.

  However, there's one thing we'd like to do, and it's not clear to us how to achieve this (or if it is even possible) in spdlog.  Basically, we're using a log with an `ostream` sink as a way to write output files from our program.  This is actually not a "log", but the main output of the tool that records data about how sequencing reads (from DNA and RNA) map to a target genome.  The output of this tool will be consumed by another tool we have written for further processing.  Because we know exactly how we want to represent each output record, it would be _much_ more efficient to do this in a binary format rather than a text-based one.  We'd still like to do this using `spdlog` as our entire output system is based around its excellent multithreaded, asynchronous logging to allow writing to file from many threads at once.  It's not clear how to do this (or if it is even possible).  I realize that this may be outside of the supported scope of spdlog, but it would be great to get your feedback / input on if there is a way to do this, and if so, what the best design might look like.

Thanks!
Rob

**Comment by gabime** [comment]
Hi,
Thanks for the nice feedback. I appreciate it!

Regarding your question, one solution the comes to mind is to define a custom logabble class:
```c++
class my_binary_record
{
private:
	const std::vector<char> _bin_data;

public:
	//copy bin data to this record
	my_binary_record(const std::vector<char>& bin_data) :_bin_data(bin_data) {};
	//or just move if possible
	my_binary_record(std::vector<char>&& bin_data) :_bin_data(move(bin_data)) {};

      //support for logging directly from spdlog
      template<typename OStream>
      friend OStream& operator<<(OStream& os, const my_binary_record &bin_record)
      {        
  		std::ostream_iterator<char> out_iter(os);		
		std::copy(bin_record._bin_data.begin(), bin_record._bin_data.end(), out_iter);		
		return os;
     }
};
```

And use it like this:
```
bin_logger->set_pattern(""); // disable formatting - will not be used
std::vector<char> some_bin_data {123,124};
my_binary_record{ std::move(some_bin_data ) };
bin_logger->info("{}", bin_record);
```

Then, define a custom sink that processes the bin data ```from message.raw``` 

```c++
class my_bin_sink : public spdlog::sinks::sink
{
  void log(const spd::details::log_msg& msg) override
  {
    const char* bin_data = msg.raw.data();
    size_t bin_size = msg.raw.size();
  }

 void flush() 
 {
   //..
 }
```

**Comment by rob-p** [comment]
Hi @gabime,

  Thanks for the quick reply; this looks like exactly what we want!  If we wanted to have this work in a multi-threaded sink, we would inherit from the base sink parameterized on a `mutex`, is that correct?  Also, the `log` method here would also, presumably, want to `.write()` the raw message to the underlying `ostream`, right?  Or is that taken care of somewhere else?  Thanks again for your help, and for the awesome library.

**Comment by gabime** [comment]
> we would inherit from the base sink parameterized on a mutex, is that correct

Right. Please note that if you plan to use only in async mode (and never share the sink between different logger objects) - no locking is needed since there is single worker thread per async logger.

>Also, the log method here would also, presumably, want to .write() the raw message to the underlying ostream, 

For what purpose ? Writing to file using ofstream?

**Comment by rob-p** [comment]
> Right. Please note that if you plan to use only in async mode (and never share the sink between different logger objects) - no locking is needed since there is single worker thread per async logger.

Perfect --- that's our intended use case.

> For what purpose ? Writing to file using ofstream?

Right.  My question likely conveys my ignorance about how `spdlog` is handling these things internally.  But what I am imagining is that this logger is associated with some `ostream` object.  That may be a file, or `stdout`, or a compressed `ostream` object (we use the `zstr` library).  Presumably, this sink should write to whatever `ostream` buffer is attached to it.  I guess my question is, is writing to the associated `ostream` object something that should happen directly i the `log` call of `my_bin_sink`?

**Comment by gabime** [comment]
> But what I am imagining is that this logger is associated with some ostream object.

Not exactly. Each sink defines and uses whatever it needs to sink its data to its destination. There is no associated ostream per logger.
So in this case your custom sink would need to write to the file by its own, by any method it chooses. (see [file sinks.h](https://github.com/gabime/spdlog/blob/master/include/spdlog/sinks/file_sinks.h) for reference).

**Comment by irowebbn** [comment]
Hi @gabime ,
I know this issue has been closed, but could you elaborate on how to create a custom formatter? The wiki documentation seems to only describe how to use set_pattern. I apologize if this question is elementary, but I am cannot find any examples of how to do this. I am trying to achieve the same thing as the OP, logging raw binary rather than text. Thank you for your help!

**Comment by rob-p** [comment]
Hi @irowebbn, 

  You can take a look at how we ended up implementing this in our tool.  It actually was quite easy.  First, we [set the logger pattern to be empty](https://github.com/COMBINE-lab/pufferfish/blob/develop/src/PufferfishAligner.cpp#L1361).  Then, we [wrote our output to a binary stream](https://github.com/COMBINE-lab/pufferfish/blob/develop/include/BinWriter.hpp#L20), which you can find the simple implementation for [here](https://github.com/COMBINE-lab/pufferfish/blob/develop/include/BinWriter.hpp#L20).  The important bit was to have the logger use an binary ostream sink, which we instantiate [here](https://github.com/COMBINE-lab/pufferfish/blob/develop/src/PufferfishAligner.cpp#L1359) and which is defined [here](https://github.com/COMBINE-lab/pufferfish/blob/develop/include/BinWriter.hpp#L111).  So, basically, you package up binary output in whichever format you want, and then you just pass it to the logger with this sink and "it just works" :).  Let me know if you have any questions.

**Comment by irowebbn** [comment]
Wow, that works perfectly! Thanks!

**Comment by dkjung** [comment]
This solution doesn't work any more, does it?

**Comment by rob-p** [comment]
Given the substantial API changes, we stuck with the previous major version of `spdlog`, so I can't comment on if this solution works in the latest version or not.  However, if it no longer does, I'd be interested in knowing why and what a possible replacement might be, as we'd eventually like to upgrade.
